### PR TITLE
Attribute mapping

### PR DESCRIPTION
**Pull request by brennovich**
Guys, I've got @ruianderson 's branch and did some work to map custom attributes with a more natural way:

```ruby
class Shot
  include Jeckle::Resource

  api :dribbble

  attribute :image_url, String, as: :image
end

# Shot.find(123).image
```

Also, this branch include work did by @tomas-stefano on `0.4.0` branch... My two cents to get the project active again :D

### CODE REVIEW

**Comment by brennovich** [comment]
I stole the ideia from `Roar`, I also tried to use `Roar` instead of `Virtus` but I had problems with mass assignment and attribute definition. `Roar` behaves more like a decorator, receiving or extending a model with already defined attributes...

File: lib/jeckle.rb [hunk 1]
```diff
@@ -2,10 +2,11 @@
 require 'faraday'
 require 'faraday_middleware'
 require 'virtus'
+require 'active_support/core_ext/array/wrap'
```
Comment (mateusg) [inline lib/jeckle.rb]: @brennovich Such a Rails person.

File: lib/jeckle/http.rb [hunk 2]
```diff
@@ -39,6 +41,23 @@ def resource_name
         @resource_name ||= model_name.element.pluralize
       end
 
+      # @public
```
Comment (mateusg) [inline lib/jeckle/http.rb]: @brennovich Do you have any intention of documenting private methods too? If not, I don't know if I would add this.
I know it's not a big deal, it's just me being annoying.

File: lib/jeckle/http.rb [hunk 3]
```diff
@@ -72,8 +92,12 @@ def api_mapping
       end
 
       def run_request(endpoint, options = {})
-        Jeckle::Request.run api_mapping[:default_api], endpoint, options
+        Jeckle::Request.run(api_mapping[:default_api], endpoint, options).tap do |request|
+          if logger = api_mapping[:default_api].logger
```
Comment (mateusg) [inline lib/jeckle/http.rb]: @brennovich What do you think about extracting `logger` to a private method?

Also, the way that it is implemented now, the assignment will always happen. Shouldn't it be `||=`?

File: lib/jeckle/http.rb [hunk 4]
```diff
@@ -72,8 +92,12 @@ def api_mapping
...
+            logger.debug("#{self} Response: #{request.response.body}")
```
Comment (mateusg) [inline lib/jeckle/http.rb]: :clap:

File: lib/jeckle/model.rb [hunk 5]
```diff
@@ -2,6 +2,7 @@ module Jeckle
   module Model
     def self.included(base)
       base.send :include, ActiveModel::Validations
+
```
Comment (mateusg) [inline lib/jeckle/model.rb]: @brennovich Who saw you, who sees you (Quem te viu, quem te vê).

File: lib/jeckle/rest_actions.rb [hunk 6]
```diff
@@ -5,19 +5,105 @@ def self.included(base)
...
+      #  @example
+      #
+      #     Post.find(1) # => posts/1
+      #
       def find(id)
-        endpoint = "#{resource_name}/#{id}"
-        attributes = run_request(endpoint).response.body
+        endpoint   = "#{resource_name}/#{id}"
+        response   = run_request(endpoint).response.body
+        attributes = parse_response(response, member_root_name)
```
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: Whaaaat?

File: lib/jeckle/rest_actions.rb [hunk 7]
```diff
@@ -5,19 +5,105 @@ def self.included(base)
...
 
-        new attributes
+        new(attributes)
```
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: Wait a minute. What happened to no-parentheses Brenno? :open_mouth:

File: lib/jeckle/rest_actions.rb [hunk 8]
```diff
@@ -5,19 +5,105 @@ def self.included(base)
...
+      #     Post.search({ status: 'published' }) # => posts/?status=published
+      #     Post.where({ status: 'published' }) # => posts/?status=published
+      #
       def search(params = {})
-        response   = run_request(resource_name, params).response.body || []
-        collection = response.kind_of?(Array) ? response : response[resource_name]
+        request = run_request(resource_name, params)
+        response = request.response
+
+        return [] unless response.success?
```
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: @brennovich Shouldn't the user get an error if the response is not success? I think this solution might hide a possible error.

Perhaps it's time to have something like `Jeckle::RequestError` with status, message and other information provided by Faraday's response. Or even create a class to encapsulate Faraday's response.

Let me know what you guys think. :bulb:

File: lib/jeckle/rest_actions.rb [hunk 9]
```diff
@@ -5,19 +5,105 @@ def self.included(base)
...
+
+        collection = parse_response(response.body, collection_root_name)
+
+        Array.wrap(collection).collect { |attrs| new attrs }
+      end
+      alias :where :search
```
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: @brennovich I know it doesn't hurt to have an alias, but shouldn't we stick with a singular interface?
(me being annoying again, I know)

File: lib/jeckle/rest_actions.rb (hunk 7 above)
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: Oh, I just realized this branch also has commits by @tomas-stefano and @ruianderson. :smile:

File: lib/jeckle/rest_actions.rb (hunk 9 above)
Comment (brennovich) [inline lib/jeckle/rest_actions.rb]: Yep, I agree :)

Not sure if we should keep this, but it was introduced with compatibility in mind. Anyway, I think we should stick with `#search`

Comment (brennovich) [inline lib/jeckle/rest_actions.rb]: Done 5267bf0

File: lib/jeckle/rest_actions.rb (hunk 8 above)
Comment (brennovich) [inline lib/jeckle/rest_actions.rb]: Not sure if we should break user's application when getting a 404. I see users doing this all the time while using `Jeckle` (override `#search` method and caring about successful responses only), I know isn't the better way but we already log requests and responses.

We can have further discussions, not sure if we should bother with this now though ;)

File: lib/jeckle/rest_actions.rb (hunk 7 above)
Comment (brennovich) [inline lib/jeckle/rest_actions.rb]: Yep! However, today I consider myself a parentheses friendly guy

File: README.md [hunk 10]
```diff
@@ -59,34 +60,29 @@ class Shot
   attribute :id, Integer
   attribute :name, String
   attribute :url, String
-  attribute :image_url, String
+  attribute :image_url, String, as: :image # You can define custom attribute mapping
```
Comment (mateusg) [inline README.md]: Maybe adding another section to explain the aliasing thing is better, isn't it?

File: lib/jeckle/http.rb (hunk 2 above)
Comment (brennovich) [inline lib/jeckle/http.rb]: After this PR I'll put some effort to normalize our docs. I really like YARD :)

File: lib/jeckle/model.rb (hunk 5 above)
Comment (brennovich) [inline lib/jeckle/model.rb]: iausdhfiusdhf I didn't get :-P

File: lib/jeckle/rest_actions.rb (hunk 6 above)
Comment (brennovich) [inline lib/jeckle/rest_actions.rb]: Don't bother with this. After `action` macro all "default" methods (find, search, etc) will be built with it, this implementation will have a short living

File: README.md (hunk 10 above)
Comment (mateusg) [inline README.md]: @brennovich I think in this case `alias` would be a more concise name for that option.
The name `as` might have an ambiguous meaning. SimpleForm and Formtastic, for example, uses `as` to determine the field type:
```ruby
f.input :age, as: :integer
```

Got it?

Comment (mateusg) [inline README.md]: To keep the same line of thought, `CustomAttributeMapping` could be named `AttributeAliasing`.

Comment (brennovich) [inline README.md]: It is just an example and we already have `attribute :image_url, String`

File: lib/jeckle/rest_actions.rb (hunk 8 above)
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: Agreed.
Maybe this behaviour should be configuration oriented.

````
config.raise_on_request_errors!
```

File: lib/jeckle/http.rb (hunk 3 above)
Comment (brennovich) [inline lib/jeckle/http.rb]: Agreed

File: lib/jeckle/http.rb (hunk 2 above)
Comment (mateusg) [inline lib/jeckle/http.rb]: :+1:

File: lib/jeckle/rest_actions.rb (hunk 6 above)
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: I was talking about adding spaces to align the `=`. This is also something Brenno wouldn't do. :stuck_out_tongue: 
And I guess it wasn't.

File: lib/jeckle/http.rb (hunk 3 above)
Comment (brennovich) [inline lib/jeckle/http.rb]: Done 825192e and 94e044a

File: README.md (hunk 10 above)
Comment (brennovich) [inline README.md]: hummm gotcha!

We may have a conceptual problem replacing `:as` with `:alias`:

```ruby
# I put attribute name passed in `:as` option as _main_ attribute
super(custom_name, coercion, options)
```
If we inspect the object we'll see `custom_name` what would't be expected with `:alias`. Also, `Roar` use `:as`, I guess is fair to compare with `Roar` than `SimpleForm`

File: lib/jeckle/rest_actions.rb (hunk 8 above)
Comment (brennovich) [inline lib/jeckle/rest_actions.rb]: Moved to #50

File: lib/jeckle/rest_actions.rb (hunk 6 above)
Comment (brennovich) [inline lib/jeckle/rest_actions.rb]: oiasdjfoisadjfoisdjf

Got it, you're I don't like it :x but it was there already...

File: lib/jeckle.rb (hunk 1 above)
Comment (brennovich) [inline lib/jeckle.rb]: :D

`Array#wrap` :heart:

File: README.md (hunk 10 above)
Comment (brennovich) [inline README.md]: Rename done 028c11a

File: lib/jeckle/rest_actions.rb (hunk 6 above)
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: I'm ok with both. :)

**Comment by mateusg** [comment]
@brennovich I think it makes sense to separate 0.4.0 related commits into #51.

I'm cleaning up this branch and leaving only things that are related to attribute mapping, so we can merge it right away.

File: lib/jeckle/model.rb (hunk 5 above)
Comment (tomas-stefano) [inline lib/jeckle/model.rb]: 'Maroto' english.

File: lib/jeckle/rest_actions.rb (hunk 6 above)
Comment (tomas-stefano) [inline lib/jeckle/rest_actions.rb]: It was me that align this. Let's remove this alignments, please. I don't do this types of things anymore.

File: lib/jeckle/rest_actions.rb (hunk 7 above)
Comment (tomas-stefano) [inline lib/jeckle/rest_actions.rb]: parenthesis

File: lib/jeckle/rest_actions.rb [hunk 11]
```diff
@@ -5,19 +5,104 @@ def self.included(base)
...
+      #     class Project
+      #       include Jeckle::Resource
+      #       api :my_api
+      #       root collection: 'awesome-projects', member: 'awesome-project'
+      #     end
+      #   end
+      #
+      def root(options={})
+        @collection_root_name = find_root_name(options[:collection], :pluralize)
+        @member_root_name     = find_root_name(options[:member], :singularize)
```
Comment (tomas-stefano) [inline lib/jeckle/rest_actions.rb]: There are alignments here too. You can remove those.

File: lib/jeckle/rest_actions.rb (hunk 6 above)
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: @tomas-stefano I think it doesn't really make much of a difference. Only if you do something like this:

```ruby
build :something, attribute_1: blablalba, attribute_2: 123123123, attribute_3: 123123123,
                            attribute_4: 123123123, attribute_5: 123123123,
```

Then it is a waste of space. One or two spaces doesn't hurt or make it any better. :)

File: lib/jeckle/rest_actions.rb (hunk 11 above)
Comment (mateusg) [inline lib/jeckle/rest_actions.rb]: @tomas-stefano I think these `root` definitions should be part of the `restful_action` DSL.

I'm working on something, will have something done soon here: #51

File: lib/jeckle/rest_actions.rb (hunk 6 above)
Comment (tomas-stefano) [inline lib/jeckle/rest_actions.rb]: Yes, agreed. But for sanity check, I will run rubocop on this repo with my personal conf. Will see the Jeckle current situation on master.

File: lib/jeckle/rest_actions.rb (hunk 11 above)
Comment (tomas-stefano) [inline lib/jeckle/rest_actions.rb]: Agreed :+1:

**Comment by mateusg** [comment]
@brennovich Done, commits removed. :shipit:

**Comment by brennovich** [comment]
:+1:

@tomas-stefano can wait the CI and merge ;)

**Comment by mateusg** [comment]
@tomas-stefano Just a sec, I'm updating README.

**Comment by brennovich** [comment]
Add `[ci skip]` to the commit msg, please

**Comment by mateusg** [comment]
@brennovich will do ;)

**Comment by mateusg** [comment]
good to go :shipit:

**Pull request merged**